jwql/shared_tasks/shared_tasks.py

```diff
@@ -0,0 +1,806 @@
+        can be spliced in to step.call() via dereferencing (**dict)
+
+    Returns
+    -------
+    reduction_path : str
+        The path at which the reduced data file(s) may be found.
+    """
+    msg = "*****CELERY: Starting {} calibration task for {}"
+    logging.info(msg.format(instrument, input_file_name))
+    config = get_config()
+    
+    input_dir = os.path.join(config['transfer_dir'], "incoming")
+    cal_dir = os.path.join(config['outputs'], "calibrated_data")
+    output_dir = os.path.join(config['transfer_dir'], "outgoing")
+    logging.info("*****CELERY: Input from {}, calibrate in {}, output to {}".format(input_dir, cal_dir, output_dir))
+    
+    input_file = os.path.join(deepcopy(input_dir), input_file_name)
+    if not os.path.isfile(input_file):
+        logging.error("*****CELERY: File {} not found!".format(input_file))
+        raise FileNotFoundError("{} not found".format(input_file))
+    
+    uncal_file = os.path.join(deepcopy(cal_dir), input_file_name)
+    ensure_dir_exists(deepcopy(cal_dir))
+    logging.info("*****CELERY: Copying {} to {}".format(input_file, cal_dir))
+    copy_files([input_file], deepcopy(cal_dir))
+    set_permissions(uncal_file)
+    
+    # Check for exposures with too many groups
+    max_groups = config.get("max_groups", 1000)
+    with fits.open(uncal_file) as inf:
+        total_groups = inf[0].header["NINTS"] * inf[0].header["NGROUPS"]
+    if total_groups > max_groups:
+        msg = "File {} has {} groups (greater than maximum of {})"
+        logging.error(msg.format(os.path.basename(uncal_file), total_groups, max_groups))
+        raise ValueError(msg.format(os.path.basename(uncal_file), total_groups, max_groups))
+    
+    steps = get_pipeline_steps(instrument)
+
+    first_step_to_be_run = True
+    for step_name in steps:
+        kwargs = {}
+        if step_name in step_args:
+            kwargs = step_args[step_name]
+        if steps[step_name]:
+            output_filename = short_name + "_{}.fits".format(step_name)
+            logging.info("*****CELERY: Output File Name is {}".format(output_filename))
+            output_file = os.path.join(cal_dir, deepcopy(output_filename))
+            logging.info("*****CELERY: Creating output file {}".format(output_file))
+            transfer_file = os.path.join(output_dir, output_filename)
+            # skip already-done steps
+            if not os.path.isfile(output_file):
+                logging.info("*****CELERY: Running Pipeline Step {}".format(step_name))
+                if first_step_to_be_run:
+                    model = PIPELINE_STEP_MAPPING[step_name].call(uncal_file, **kwargs)
+                    first_step_to_be_run = False
+                else:
+                    model = PIPELINE_STEP_MAPPING[step_name].call(model, **kwargs)
+
+                if step_name != 'rate':
+                    # Make sure the dither_points metadata entry is at integer (was a
```

Comment:
Okay. I guess the question is whether we might still be getting uncalibrated files that were produced a long time ago, and that need to be converted to work in the current pipeline. I know that on the HST side, it tended to be that the "raw" data would accumulate sets of things that you needed to look out for depending on how it was generated, until eventually the DB/pipeline team (whose name at the time I forget now) would do a full reprocessing of everything to match what the current pipeline expected, and then the process would repeat.

As far as I recall, when we do pipeline processing we're *always* grabbing the data from the Isilon filesystem, so I guess the question is whether the _uncal.fits files stored there ever change, or whether they just stay as they were created forever.

Reply:
This is a good point @bhilbert4 do you know if this will cause any issues? Maybe we can ask someone in the engineering divisions if they update the isilon data when they do any sort of pipeline updating.

Reply:
JWST data are periodically reprocessed when important new reference files are delivered or a pipeline step is updated, etc. That reprocessing can, but does not have to, include re-creation of the uncal files. But more importantly, the pipeline update that this code is addressing was made well before launch. There were 6 or 7 pipeline releases between that update and launch. So there should not be any flight data that have the old type. So I think it's safe to remove. I'm agnostic about whether we remove that here, or wait and do it down the line in a separate PR, since it isn't hurting anything by having the lines in here.

Reply:
I think a separate PR makes sense. It seems like (as with another comment thread) this might make a good introductory issue, in that it might also include talking to a few of the people who are involved in JWST ops, which also can't hurt.